### PR TITLE
Unary operators and UDAF support for DECIMAL fields

### PR DESCRIPTION
**Pull request by agavra**
### Description 

* Support unary `-` and `+` for decimal columns
* Support for SUM, MIN, MAX and TopKDistinct

### Testing done 
* New unit tests
* New QTT tests

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-serde/src/main/java/io/confluent/ksql/serde/delimited/KsqlDelimitedSerializer.java [hunk 1]
```diff
@@ -83,7 +84,20 @@ public Object next() {
       final Field field = fieldIt.next();
       throwOnUnsupportedType(field.schema());
       if (DecimalUtil.isDecimal(field.schema())) {
-        return ((BigDecimal) data.get(field)).toPlainString();
+        final BigDecimal value = (BigDecimal) data.get(field);
```
Comment (hjafarpour) [inline ksql-serde/src/main/java/io/confluent/ksql/serde/delimited/KsqlDelimitedSerializer.java]: Adding a test for this would be great! :)

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/CodeGenRunner.java [hunk 2]
```diff
@@ -211,6 +212,13 @@ protected Object visitArithmeticBinary(
       return null;
     }
 
+    protected Object visitArithmeticUnary(
```
Comment (vcrfxia) [inline ksql-engine/src/main/java/io/confluent/ksql/codegen/CodeGenRunner.java]: How come this method didn't exist previously? Is this fixing a bug or is it only needed for decimals?

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java [hunk 3]
```diff
@@ -456,11 +456,31 @@ private String visitBooleanComparisonExpression(final ComparisonExpression.Type
...
         case MINUS:
-          // this is to avoid turning a sequence of "-" into a comment (i.e., "-- comment")
-          final String separator = value.getLeft().startsWith("-") ? " " : "";
-          return new Pair<>("-" + separator + value.getLeft(), value.getRight());
+          if (DecimalUtil.isDecimal(value.getRight())) {
+            return new Pair<>(
+                String.format(
+                    "(%s.negate(new MathContext(%d, RoundingMode.UNNECESSARY)))",
+                    value.getLeft(),
+                    DecimalUtil.precision(value.getRight())),
```
Comment (vcrfxia) [inline ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: nit: unrelated to this PR, but reading the implementation for `DecimalUtil.precision()` is confusing since the intermediate variable is called `scaleString`.

File: ksql-functional-tests/src/test/resources/query-validation-tests/min-group-by.json [hunk 4]
```diff
@@ -134,6 +134,49 @@
...
+        {"topic": "S2", "key": 100, "value": "100,06.40"},
+        {"topic": "S2", "key": 0, "value": "0,-01.79"},
+        {"topic": "S2", "key": 0, "value": "0,-01.79"}
+      ]
+    },
+    {
+      "name": "min decimal group by with nulls",
+      "statements": [
+        "CREATE STREAM TEST (ID bigint, VALUE decimal(4, 2)) WITH (kafka_topic='test_topic',value_format='AVRO', key='ID');",
+        "CREATE TABLE S2 as SELECT id, min(value) as MAX FROM test group by id;"
```
Comment (vcrfxia) [inline ksql-functional-tests/src/test/resources/query-validation-tests/min-group-by.json]: nit: `min(value) as MIN`

File: ksql-serde/src/test/java/io/confluent/ksql/serde/delimited/KsqlDelimitedDeserializerTest.java [hunk 5]
```diff
@@ -185,6 +190,26 @@ public void shouldDeserializedTopLevelPrimitiveTypeIfSchemaHasOnlySingleField()
...
+    // Given:
+    final PersistenceSchema schema = persistenceSchema(
+        SchemaBuilder.struct()
+          .field("cost", DecimalUtil.builder(4, 2))
+          .build()
+    );
+    final KsqlDelimitedDeserializer deserializer =
+        new KsqlDelimitedDeserializer(schema, recordLogger);
+
+    final byte[] bytes = "01.12".getBytes(StandardCharsets.UTF_8);
```
Comment (vcrfxia) [inline ksql-serde/src/test/java/io/confluent/ksql/serde/delimited/KsqlDelimitedDeserializerTest.java]: We'd also correctly deserialize if the leading 0 were missing, right? Perhaps worth adding a test to ensure that.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topkdistinct/TopkDistinctAggFunctionFactory.java [hunk 6]
```diff
@@ -18,59 +18,50 @@
...
-      throw new KsqlException("No TOPKDISTINCT aggregate function with " + argTypeList.get(0)
-          + " argument type exists!");
+    final Schema argSchema = argTypeList.get(0);
+    switch (argSchema.type()) {
+      case INT32:
+      case INT64:
+      case FLOAT64:
+      case STRING:
+        return new TopkDistinctKudaf(NAME, 0, -1, argSchema, SchemaUtil.getJavaType(argSchema));
+      case BYTES:
```
Comment (vcrfxia) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topkdistinct/TopkDistinctAggFunctionFactory.java]: I see you've added query validation tests for this, but perhaps also worth adding a `DecimalTopkDistinctKudafTest.java` unit test file, analogous to the existing ones for the other types?

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/CodeGenRunner.java (hunk 2 above)
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/codegen/CodeGenRunner.java]: yes - previously all numerics were terminal primitive types. that is no longer necessarily the case

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java (hunk 3 above)
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: oops! I'll fix that

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topkdistinct/TopkDistinctAggFunctionFactory.java (hunk 6 above)
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topkdistinct/TopkDistinctAggFunctionFactory.java]: :scratches-head: I was sure I had done that!

File: ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java [hunk 7]
```diff
@@ -500,11 +500,31 @@ private String visitBooleanComparisonExpression(final ComparisonExpression.Type
       final Pair<String, Schema> value = process(node.getValue(), context);
       switch (node.getSign()) {
         case MINUS:
-          // this is to avoid turning a sequence of "-" into a comment (i.e., "-- comment")
-          final String separator = value.getLeft().startsWith("-") ? " " : "";
-          return new Pair<>("-" + separator + value.getLeft(), value.getRight());
+          if (DecimalUtil.isDecimal(value.getRight())) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java]: This method is getting larger with this enhancement. Consider refactoring. 

You could either refactor each case into a method, e.g.

```java
switch (node.getSign()) {
case Minus:  return visitArithmeticMinus(...);
case Plus: return visitArithmeticPlus(...);
}
```

Or pull out the decimal as a special case, e.g.

```java
if (DecimalUtil.isDecimal(value.getRight())) {
   return visitDecimalArithmeticUnary(..);
}

// old code here
}
```

Personally, I prefer the first.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/sum/DecimalSumKudaf.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+import io.confluent.ksql.function.KsqlAggregateFunction;
+import io.confluent.ksql.function.TableAggregationFunction;
+import io.confluent.ksql.util.DecimalUtil;
+import java.math.BigDecimal;
+import java.math.MathContext;
+import java.util.Collections;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.streams.kstream.Merger;
+
+public class DecimalSumKudaf
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/sum/DecimalSumKudaf.java]: Weird arse class. But fixing this is outside the scope of this PR...

File: ksql-functional-tests/src/test/resources/query-validation-tests/max-group-by.json [hunk 9]
```diff
@@ -136,6 +136,49 @@
...
+        {"topic": "S2", "key": 0, "value": "0,05.10"},
+        {"topic": "S2", "key": 100, "value": "100,10.10"},
+        {"topic": "S2", "key": 100, "value": "100,10.10"},
+        {"topic": "S2", "key": 100, "value": "100,30.80"},
+        {"topic": "S2", "key": 0, "value": "0,20.99"},
+        {"topic": "S2", "key": 0, "value": "0,20.99"}
+      ]
+    },
+    {
+      "name": "max decimal by with nulls",
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/max-group-by.json]: I'd personally roll this into the test above, so the one test covers null and non-null.

File: ksql-functional-tests/src/test/resources/query-validation-tests/min-group-by.json [hunk 10]
```diff
@@ -134,6 +134,49 @@
...
+        {"topic": "S2", "key": 0, "value": "0,01.10"},
+        {"topic": "S2", "key": 100, "value": "100,10.10"},
+        {"topic": "S2", "key": 100, "value": "100,06.40"},
+        {"topic": "S2", "key": 100, "value": "100,06.40"},
+        {"topic": "S2", "key": 0, "value": "0,-01.79"},
+        {"topic": "S2", "key": 0, "value": "0,-01.79"}
+      ]
+    },
+    {
+      "name": "min decimal group by with nulls",
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/min-group-by.json]: likewise, consider rolling into the above test.

File: ksql-functional-tests/src/test/resources/query-validation-tests/max-group-by.json [hunk 11]
```diff
@@ -136,6 +136,49 @@
...
+      "statements": [
+        "CREATE STREAM TEST (ID bigint, VALUE decimal(4,2)) WITH (kafka_topic='test_topic',value_format='AVRO', key='ID');",
+        "CREATE TABLE S2 as SELECT id, max(value) as MAX FROM test group by id;"
+      ],
+      "inputs": [
+        {"topic": "test_topic", "key": 1, "value": {"id":  1, "value":  null}},
+        {"topic": "test_topic", "key": 0, "value": {"id":  0, "value":  null}},
+        {"topic": "test_topic", "key": 0, "value": {"id":  0, "value":  "01.00"}}
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/max-group-by.json]: be good to test what happens if `null` value seen after a valid value.

File: ksql-functional-tests/src/test/resources/query-validation-tests/min-group-by.json [hunk 12]
```diff
@@ -134,6 +134,49 @@
...
+      "statements": [
+        "CREATE STREAM TEST (ID bigint, VALUE decimal(4, 2)) WITH (kafka_topic='test_topic',value_format='AVRO', key='ID');",
+        "CREATE TABLE S2 as SELECT id, min(value) as MIN FROM test group by id;"
+      ],
+      "inputs": [
+        {"topic": "test_topic", "key": 1, "value": {"id":  1, "value":  null}},
+        {"topic": "test_topic", "key": 0, "value": {"id":  0, "value":  null}},
+        {"topic": "test_topic", "key": 0, "value": {"id":  0, "value":  "01.00"}}
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/min-group-by.json]: be good to test what happens if `null` value seen after a valid value.

File: ksql-functional-tests/src/test/resources/query-validation-tests/sum.json [hunk 13]
```diff
@@ -89,6 +89,25 @@
...
+      "name": "sum decimal",
+      "statements": [
+        "CREATE STREAM TEST (ID bigint, NAME varchar, VALUE decimal(4,1)) WITH (kafka_topic='test_topic', value_format='DELIMITED', key='ID');",
+        "CREATE TABLE S2 as SELECT id, sum(value) FROM test group by id;"
+      ],
+      "inputs": [
+        {"topic": "test_topic", "key": 0, "value": "0,zero,5.4"},
+        {"topic": "test_topic", "key": 0, "value": "0,100,100.1"},
+        {"topic": "test_topic", "key": 100, "value": "100,100,500.9"},
+        {"topic": "test_topic", "key": 100, "value": "100,100,300.8"}
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/sum.json]: Add in a null value example.

File: ksql-functional-tests/src/test/resources/query-validation-tests/sum.json [hunk 14]
```diff
@@ -89,6 +89,25 @@
         {"topic": "S2", "key": 100, "value": "100,801.7"}
       ]
     },
+    {
+      "name": "sum decimal",
+      "statements": [
+        "CREATE STREAM TEST (ID bigint, NAME varchar, VALUE decimal(4,1)) WITH (kafka_topic='test_topic', value_format='DELIMITED', key='ID');",
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/sum.json]: nit: remove NAME column as its just noise in the test.

File: ksql-functional-tests/src/test/resources/query-validation-tests/max-group-by.json [hunk 15]
```diff
@@ -136,6 +136,49 @@
         {"topic": "S2", "key": 0, "value": {"ID":  0, "MAX":  null}},
         {"topic": "S2", "key": 0, "value": {"ID":  0, "MAX":  1.0}}
       ]
+    },
+    {
+      "name": "max decimal group by",
+      "statements": [
+        "CREATE STREAM TEST (ID bigint, NAME varchar, VALUE decimal(4, 2)) WITH (kafka_topic='test_topic', value_format='DELIMITED', key='ID');",
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/max-group-by.json]: nit: remove NAME column as its just noise in the test.

File: ksql-functional-tests/src/test/resources/query-validation-tests/min-group-by.json [hunk 16]
```diff
@@ -134,6 +134,49 @@
         {"topic": "S2", "key": 0, "value": {"ID":  0, "MAX":  null}},
         {"topic": "S2", "key": 0, "value": {"ID":  0, "MAX":  1.0}}
       ]
+    },
+    {
+      "name": "min decimal group by",
+      "statements": [
+        "CREATE STREAM TEST (ID bigint, NAME varchar, VALUE decimal(4, 2)) WITH (kafka_topic='test_topic', value_format='DELIMITED', key='ID');",
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/min-group-by.json]: nit: remove NAME column as its just noise in the test.

File: ksql-functional-tests/src/test/resources/query-validation-tests/topk-distinct.json [hunk 17]
```diff
@@ -70,6 +70,27 @@
         {"topic": "S2", "value": {"ID":0,"TOPK":["c","b","a"]}},
         {"topic": "S2", "value": {"ID":0,"TOPK":["d","c","b"]}}
       ]
+    },
+    {
+      "name": "topk distinct decimal",
+      "statements": [
+        "CREATE STREAM TEST (ID bigint, NAME varchar, VALUE decimal(2,1)) WITH (kafka_topic='test_topic', value_format='AVRO', key='ID');",
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/topk-distinct.json]: nit: remove NAME column as its just noise in the test.

File: ksql-functional-tests/src/test/resources/query-validation-tests/topk-distinct.json [hunk 18]
```diff
@@ -70,6 +70,27 @@
...
+        "CREATE TABLE S2 as SELECT id, topkdistinct(value, 3) as topk FROM test group by id;"
+      ],
+      "inputs": [
+        {"topic": "test_topic", "value": {"ID": 0, "NAME": "zero", "VALUE": "9.8"}},
+        {"topic": "test_topic", "value": {"ID": 0, "NAME": "zero", "VALUE": "8.9"}},
+        {"topic": "test_topic", "value": {"ID": 0, "NAME": "zero", "VALUE": "7.8"}},
+        {"topic": "test_topic", "value": {"ID": 0, "NAME": "zero", "VALUE": "6.5"}},
+        {"topic": "test_topic", "value": {"ID": 0, "NAME": "zero", "VALUE": "9.9"}}
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/topk-distinct.json]: add a null value example, early on.

File: ksql-serde/src/main/java/io/confluent/ksql/serde/delimited/KsqlDelimitedSerializer.java [hunk 19]
```diff
@@ -83,7 +84,15 @@ public Object next() {
...
```
Comment (big-andy-coates) [inline ksql-serde/src/main/java/io/confluent/ksql/serde/delimited/KsqlDelimitedSerializer.java]: I'd move this into its own method.

```java
if (decimal) {
   return getDecimal(field);
}

return format.format(value);
```

File: ksql-common/src/main/java/io/confluent/ksql/function/AggregateFunctionFactory.java [hunk 20]
```diff
@@ -33,6 +34,7 @@
       .add(ImmutableList.of(Schema.OPTIONAL_INT32_SCHEMA))
       .add(ImmutableList.of(Schema.OPTIONAL_INT64_SCHEMA))
       .add(ImmutableList.of(Schema.OPTIONAL_FLOAT64_SCHEMA))
+      .add(ImmutableList.of(DecimalUtil.builder(1, 1).build()))
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-common/src/main/java/io/confluent/ksql/function/AggregateFunctionFactory.java]: yes - though this is just for use in describe function. I couldn't really think of a better way to do this (also from the UI perspective)

File: ksql-functional-tests/src/test/resources/query-validation-tests/max-group-by.json (hunk 9 above)
Comment (agavra) [reply ksql-functional-tests/src/test/resources/query-validation-tests/max-group-by.json]: I'll have a minor follow-up to do this because it's just kinda annoying and I want to get this in :)

File: ksql-functional-tests/src/test/resources/query-validation-tests/sum.json (hunk 14 above)
Comment (agavra) [reply ksql-functional-tests/src/test/resources/query-validation-tests/sum.json]: I will do this in the same follow-up as above since and just change the whole file

File: ksql-functional-tests/src/test/resources/query-validation-tests/sum.json (hunk 13 above)
Comment (agavra) [reply ksql-functional-tests/src/test/resources/query-validation-tests/sum.json]: ditto - will do this in follow-up PR right after submitting this (will require a format change, will change all hte tests)
